docs/api-change-runbook.md

```diff
@@ -0,0 +1,264 @@
+* If you need to redirect an endpoint, you have to clone the old one under the
+  new URI in the swagger specification.
+
+### Marking endpoints as deprecated
+
+When marking:
+
+* an HTTP endpoint as deprecated:
+  * Add a comment for the parsing function of the endpoint stating that it
+    is deprecated.
+  * Log a `warn!` message stating that the user accessed a deprecated
+    endpoint.
+  * Increment the `deprecatedHttpApi` metric.
+  * Include the `Deprecated` header in the response.
+* a header field in an HTTP endpoint as deprecated:
+  * Add a comment in the parsing function where we check the presence of the
+    header stating that it is deprecated.
+  * If the header is present, log a `warn!` message stating that the user
+    used a deprecated field.
+  * Increment the `deprecatedHttpApi` metric.
+  * Include the Deprecated header in the response.
+* a command line parameter as deprecated:
+  * Mention it is deprecated in the help message of the parameter in the
+    argument parser.
+  * Add it in the `warn_deprecated_parameters` function where we log it
+    and increment the `deprecatedCmdLineApi` metric.
+
+## Practical example of an API change
+
+In this guide we set out to remove the `vsock_id` field in `PUT`s on `/vsock`.
+This was implemented in [PR #2763][2] and we will go step by step through the
+changes in order to understand the process of changing something in the
+Firecracker API.
+
+* We go through the flowchart; we want to remove a field in the body of a
+  HTTP request. So we follow the flowchart like this:
+  * → Change an existing endpoint
+  * → Request
+  * → Remove header or field
+  * → Make it optional
+  * → Deprecate
+  * → Increment minor version.
+* Now that we know we need to make the field optional and deprecate it, it’s
+  time for the code changes (reference implementation in [this commit][3]).
+  We go to the function in `api_server/src/requests` which is responsible for
+  parsing this request, which is `parse_put_vsock` in this case, and do the
+  following.
+  * We find the associated `vmm_config` struct which `serde_json` uses for
+    deserialization, in this case `VsockDeviceConfig`.
+  * In the struct referenced above, we make the parameter optional by
+     encapsulating it in an `Option` with `#[serde(default)]` and
+     `#[serde(skip_serializing_if = "Option::is_none")]` so that we don’t break
+     existing implementations, but we follow the new, desired usage of the
+     endpoint.
+  * After deserializing the body of the request into the struct, we check
+     for the existence of the field we want to deprecate, in this case by
+     calling `vsock_cfg.vsock_id.is_some()`.
+  * If the field is there, we must mark this request as being deprecated,
+     so we craft a deprecation message
+     (`"PUT /vsock: vsock_id field is deprecated."`) and increment the
+     deprecated HTTP API metric
+     (`METRICS.deprecated_api.deprecated_http_api_calls.inc()`).
```

Comment:
Not necessarily related to this doc, but how do we make sure this is updated when the deprecated entity is no longer under a deprecated state?

I guess the bigger question is, what s the mechanism for coming out of a deprecation state?

Reply:
This metric tracks the number of API calls made by users on endpoints which were either entirely deprecated or used fields which were deprecated. If, for the lifetime of the VM, users never call any deprecated endpoints, the metric stays at 0. This is just a tool for users to find out through the metrics if they are using any deprecated endpoints and how often.

Also added a small section for removing deprecated elements on a major version release,